### PR TITLE
[headless-cache-tuning] Patch 4: Scale --max-turns with patch complexity

### PR DESCRIPTION
**Pull request by subsetpark**
## Patch 4: Scale --max-turns with patch complexity

- Add `let max_turns_for ~complexity = match complexity with Some 1 -> 50 | Some 2 -> 100 | _ -> 200` (Some 3 and None fall through to 200).
- Thread `complexity` into build_args / build_stream_args so they can compute the cap. (build_stream_args already receives complexity via run_streaming's resolve_auto_model call — extend the build_*_args signature similarly.)
- Update existing inline tests that assert exact argv to use the new variable cap.
- Add inline tests for max_turns_for: 1 -> 50, 2 -> 100, 3 -> 200, None -> 200.

## Changes
- Add `let max_turns_for ~complexity = match complexity with Some 1 -> 50 | Some 2 -> 100 | _ -> 200` (Some 3 and None fall through to 200).
- Thread `complexity` into build_args / build_stream_args so they can compute the cap. (build_stream_args already receives complexity via run_streaming's resolve_auto_model call — extend the build_*_args signature similarly.)
- Update existing inline tests that assert exact argv to use the new variable cap.
- Add inline tests for max_turns_for: 1 -> 50, 2 -> 100, 3 -> 200, None -> 200.

## Files to Modify
- lib/claude_runner.ml (modify): Add `max_turns_for ~complexity` returning 50/100/200 for complexity 1/2/3 (None falls to 200). Replace the hard-coded 200 in build_args and build_stream_args.

## Gameplan Specification

```
module HEADLESS_CACHE_TUNING.

> ══════════════════════════════════════════
> THE GUARANTEE
> ══════════════════════════════════════════
> After this gameplan, every onton-spawned coding-agent
> invocation is tuned for prompt-cache reuse and bounded
> token consumption. Static prompt prefixes are reused
> across patches in one gameplan. Isolated per-spawn
> config dirs stabilise spawn environment. Onton-minted
> session UUIDs remove the parsing race and enable resume
> without re-sending the prompt. Turn caps scale with
> complexity. Per-spawn USD budget caps put a hard floor
> under runaway token spend.

Patch.
Gameplan.
Spawn.
Arg.
Prompt.
Text.
SessionId.
Dir.
Backend.

claude => Backend.
codex => Backend.
pi => Backend.

backend s: Spawn => Backend.
spawn-of p: Patch => Spawn.
complexity p: Patch => Nat.

render p: Patch, gp: Gameplan => Prompt.
prefix pr: Prompt => Text.
suffix pr: Prompt => Text.
shared? t: Text, gp: Gameplan => Bool.
specific? t: Text, p: Patch => Bool.

args s: Spawn => Arg.
has-flag? a: Arg, name: String => Bool.

config-dir s: Spawn => Dir.
max-turns s: Spawn => Nat.

session-id p: Patch => SessionId.
persisted-before-spawn? id: SessionId, s: Spawn => Bool.
passed-to-cli? id: SessionId, s: Spawn => Bool.
minted-session-flag-on? => Bool.

budget-cap-set? s: Spawn => Bool.
flag-emitted? s: Spawn => Bool.
cost-tracked? s: Spawn => Bool.
self-terminates-on-overshoot? s: Spawn => Bool.

bare-flag-on? => Bool.
worktree-claude-md p: Patch => Text.
non-empty? t: Text => Bool.
prompt-of s: Spawn => Prompt.
contains? t: Text, sub: String => Bool.
---

> Static prefix is gameplan-shared.
all p: Patch, gp: Gameplan | shared? (prefix (render p gp)) gp.

> Dynamic suffix is patch-specific.
all p: Patch, gp: Gameplan | specific? (suffix (render p gp)) p.

> Claude spawns pass --exclude-dynamic-system-prompt-sections.
all s: Spawn, backend s = claude |
  has-flag? (args s) "--exclude-dynamic-system-prompt-sections".

> Distinct patches have distinct config dirs.
all p1: Patch, p2: Patch, p1 ~= p2 |
  config-dir (spawn-of p1) ~= config-dir (spawn-of p2).

> Turn caps are monotone non-decreasing in complexity.
all p1: Patch, p2: Patch, complexity p1 < complexity p2 |
  max-turns (spawn-of p1) <= max-turns (spawn-of p2).

> Session-ID minting (when enabled) persists before spawn and
> passes to the CLI.
all p: Patch, minted-session-flag-on? |
  persisted-before-spawn? (session-id p) (spawn-of p).

all p: Patch, minted-session-flag-on? |
  passed-to-cli? (session-id p) (spawn-of p).

> When the budget cap is set, Claude spawns emit the flag and
> Codex spawns track cost + self-terminate.
all s: Spawn, backend s = claude, budget-cap-set? s |
  flag-emitted? s.

all s: Spawn, backend s = codex, budget-cap-set? s |
  cost-tracked? s and self-terminates-on-overshoot? s.

> When --bare is enabled, Claude spawns pass --bare and the
> worktree's CLAUDE.md content is reachable from the spawn's
> prompt prefix.
all p: Patch, bare-flag-on?, backend (spawn-of p) = claude |
  has-flag? (args (spawn-of p)) "--bare".

all p: Patch, bare-flag-on?, non-empty? (worktree-claude-md p) |
  contains? (prefix (prompt-of (spawn-of p))) "CLAUDE.md".

```

## Patch Specification

```
module HEADLESS_CACHE_TUNING_PATCH_4.

> Patch 5 scales --max-turns with patch complexity so simple
> patches can't burn the full 200-turn budget.

Patch.
Spawn.

complexity p: Patch => Nat.
spawn-of p: Patch => Spawn.
max-turns s: Spawn => Nat.
---

> Turn cap is monotone non-decreasing in complexity.
all p1: Patch, p2: Patch, complexity p1 < complexity p2 |
  max-turns (spawn-of p1) <= max-turns (spawn-of p2).

```

## Implementation Notes

- The new cap is computed at the Claude runner boundary and threaded through both `run` and `run_streaming` so the helper stays the single source of truth for argv generation.
- I kept the ladder intentionally coarse: complexity 3 and unknown/absent complexity both stay on 200, matching the patch’s stated policy and preserving existing fallback behavior.

## Release Notes

* **New Features**
  * Added optional complexity parameter to dynamically adjust AI processing capacity based on task difficulty (lower complexity enables faster execution, higher complexity enables deeper analysis)

* **Tests**
  * Expanded test coverage for new complexity-based functionality

### CODE REVIEW

File: lib/claude_runner.ml [hunk 1]
```diff
@@ -296,9 +311,24 @@ let%test "auto_model: None -> opus (conservative)" =
...
+let%test "max_turns_for: complexity 2 -> 100" =
+  Int.equal (max_turns_for ~complexity:(Some 2)) 100
+
+let%test "max_turns_for: complexity 3 -> 200" =
+  Int.equal (max_turns_for ~complexity:(Some 3)) 200
+
+let%test "max_turns_for: None -> 200" =
+  Int.equal (max_turns_for ~complexity:None) 200
+
 let%test "build_args fresh (no resume, with model)" =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/claude_runner.ml:331]: Fixed in the follow-up: the argv tests now use literal turn caps, so they independently validate `max_turns_for` instead of reusing its output.

File: lib/claude_runner.ml [hunk 2]
```diff
@@ -63,13 +66,12 @@ let build_args ~model ~prompt ~resume_session =
     [
       "--dangerously-skip-permissions";
       "--max-turns";
-      "200";
-      "--exclude-dynamic-system-prompt-sections";
+      Int.to_string (max_turns_for ~complexity);
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/claude_runner.ml:69]: This is already addressed at `HEAD`: the Claude runner keeps `--exclude-dynamic-system-prompt-sections` in both argv builders, and the membership tests are present again. The comment is outdated after the rebase.
